src/main/kotlin/org/rust/lang/core/CFGBuilder.kt

```
@@ -0,0 +1,366 @@
+        val newNode = graph.addNode(data)
+        preds.forEach { addContainedEdge(it, newNode) }
+        return newNode
+    }
+
+    fun addContainedEdge(source: CFGIndex, target: CFGIndex) {
+        val data = CFGEdgeData(mutableListOf())
+        graph.addEdge(source, target, data)
+    }
+
+    private fun addReturningEdge(fromIndex: CFGIndex) {
+        val data = CFGEdgeData(loopScopes.map { it.loop }.toMutableList())
+        graph.addEdge(fromIndex, exit, data)
+    }
+
+    private fun straightLine(expr: RsExpr, pred: CFGIndex, subExprs: List<RsExpr?>): CFGIndex {
+        val subExprsExit = subExprs.fold(pred) { pred, subExpr -> process(subExpr, pred) }
+        return addAstNode(expr, listOf(subExprsExit))
+    }
+
+    // This function will be used during break/continue expressions processing
+    // TODO: change return type to Pair<Region.Scope, NodeIndex>
+    private fun findScopeEdge(expr: RsExpr, destination: Destination, kind: ScopeControlFlowKind): CFGIndex? {
+        for (b in breakableBlockScopes) {
+            if (b.block == destination.target) {
+                return when (kind) {
+                    ScopeControlFlowKind.BREAK -> b.breakIndex
+                    ScopeControlFlowKind.CONTINUE -> null
+                }
+            }
+        }
+        for (l in loopScopes) {
+            if (l.loop == destination.target) {
+                return when (kind) {
+                    ScopeControlFlowKind.BREAK -> l.breakIndex
+                    ScopeControlFlowKind.CONTINUE -> l.continueIndex
+                }
+            }
+        }
+
+        return null
+    }
+
+    fun process(element: RsElement?, pred: CFGIndex): CFGIndex {
+        if (element == null) return pred
+
+        val oldPredsSize = preds.size
+        preds.push(pred)
+        element.accept(this)
+        preds.pop()
+        assert(preds.size == oldPredsSize)
+
+        return result!!
```

Maybe It's better to use `checkNotNull`:
```kotlin
return checkNotNull(result) { "Optional (lazy) error message" }
```
This method throws `IllegalStateException` instead of `NullPointerException`.